[LLD][COFF] Process all live import symbols in MapFile's getSymbols()

cjacek opened the pull request:
The current logic assumes that the import file is pulled by object files, and the loop for import files only needs to handle cases where the `__imp_` symbol is implicitly pulled by an import thunk. This is fragile, as the symbol may also be pulled through other means, such as the -export argument in tests. Additionally, this logic is insufficient for ARM64EC, which exposes multiple symbols through an import file, and referencing any one of them causes all of them to be defined.
    
With this change, import symbols are added to `syms` more often, but we ensure that output symbols remain unique later in the process

cjacek commented:
Depends on #109115.

mstorsjo commented on lld/COFF/MapFile.cpp:
```diff
@@ -122,16 +122,9 @@ static void getSymbols(const COFFLinkerContext &ctx,
-
-    syms.push_back(thunkSym);
-
-    if (auto *impSym = dyn_cast_or_null<Defined>(file->impSym))
```
So technically, it's not possible for `impSym` to be null, or to be anything else than a `Defined` here?

mstorsjo commented on lld/COFF/MapFile.cpp:
```diff
@@ -122,16 +122,9 @@ static void getSymbols(const COFFLinkerContext &ctx,
-
-    if (auto *impSym = dyn_cast_or_null<Defined>(file->impSym))
-      syms.push_back(impSym);
+    syms.push_back(file->impSym);
```
I asked about this in https://github.com/llvm/llvm-project/pull/108459#discussion_r1758748174 before, but I didn't quite grasp the response then, but now I understand.

So if we have another object file referencing e.g. `__imp_func` (which is an undefined reference from the scope of that object file, originally), that defined symbol will get picked up and added to the list of symbols to include in the loop above, looping over `ctx.objFileInstances`.

That's why a regular reference to `__imp_func`, without referencing `func` at all, would have been listed in the map file just fine. But if we pull in `__imp_func` via an `-export:` option, it isn't so we need to include it here. Now I get it!

mstorsjo reviewed (approved):
LGTM

mstorsjo commented:
Oh btw, please update the patch/PR subject to mention MapFile in some form - `getSymbols()` on its own doesn't really say much about the scope/context about what it does.

cjacek committed:
[LLD][COFF] Process all live import symbols in MapFile's getSymbols()

The current logic assumes that the import file is pulled by object files,
and the loop for import files only needs to handle cases where the __imp_
symbol is implicitly pulled by an import thunk. This is fragile, as the
symbol may also be pulled through other means, such as the -export argument
in tests. Additionally, this logic is insufficient for ARM64EC, which
exposes multiple symbols through an import file, and referencing any one
of them causes all of them to be defined.

With this change, import symbols are added to syms more often, but we
ensure that output symbols remain unique later in the process.
